src/Bicep.Core/Parsing/Parser.cs

```diff
@@ -1,370 +1,184 @@
-
-                    return current.Type switch
-                    {
-                        TokenType.Identifier => current.Text switch
-                        {
-                            LanguageConstants.TargetScopeKeyword => this.TargetScope(),
-                            LanguageConstants.ParameterKeyword => this.ParameterDeclaration(),
-                            LanguageConstants.VariableKeyword => this.VariableDeclaration(),
-                            LanguageConstants.ResourceKeyword => this.ResourceDeclaration(),
-                            LanguageConstants.OutputKeyword => this.OutputDeclaration(),
-                            LanguageConstants.ModuleKeyword => this.ModuleDeclaration(),
-                            _ => throw new ExpectedTokenException(current, b => b.UnrecognizedDeclaration()),
-                        },
-                        TokenType.NewLine => this.NewLine(),
-
-                        _ => throw new ExpectedTokenException(current, b => b.UnrecognizedDeclaration()),
-                    };
-                },
-                RecoveryFlags.None,
-                TokenType.NewLine);
-
-        private static RecoveryFlags GetSuppressionFlag(SyntaxBase precedingNode)
+                        while (true)
+                        {
+                            if (this.Check(TokenType.At))
+                            {
+                                leadingNodes.Add(this.Decorator());
+                                continue;
+                            }
+
+                            if (this.Check(TokenType.NewLine))
+                            {
+                                var newLine = this.WithRecoveryNullable(this.NewLineOrEof, RecoveryFlags.ConsumeTerminator, TokenType.NewLine);
+
+                                if (newLine != null)
+                                {
+                                    leadingNodes.Add(newLine);
+                                }
+
+                                continue;
+                            }
+
+                            break;
+                        }
+                    }
+
+                    Token current = reader.Peek();
+
+                    return current.Type switch
+                    {
+                        TokenType.Identifier => current.Text switch
+                        {
+                            LanguageConstants.TargetScopeKeyword => this.TargetScope(leadingNodes),
+                            LanguageConstants.ParameterKeyword => this.ParameterDeclaration(leadingNodes),
+                            LanguageConstants.VariableKeyword => this.VariableDeclaration(leadingNodes),
+                            LanguageConstants.ResourceKeyword => this.ResourceDeclaration(leadingNodes),
+                            LanguageConstants.OutputKeyword => this.OutputDeclaration(leadingNodes),
+                            LanguageConstants.ModuleKeyword => this.ModuleDeclaration(leadingNodes),
```

Comment:
Out of interest, what are the pros/cons of modelling decorators as part of the declaration syntax, vs as a top-level syntax independent of the declarations? Do you happen to know how this is handled in similar languages (C# or TS for example)?

Reply:
Roslyn includes under the declaration:
![image](https://user-images.githubusercontent.com/22460039/105764152-765cbc00-5f0b-11eb-920d-4077fe7ecabc.png)

For the missing declaration case, they create an IncompleteMember node:
![image](https://user-images.githubusercontent.com/22460039/105764414-c471bf80-5f0b-11eb-98ea-95790d1976db.png)

Reply:
I suspect having the decorators under the declaration makes it easier to process them. Otherwise, you would always need a visitor to collect the preceding decorators. (Or the ability to efficiently get the preceding peer node in the AST.)

Reply:
TypeScript also put them under declarations probably for the reason that Marcin mentioned:
![image](https://user-images.githubusercontent.com/16367959/105765506-6645dc00-5f0d-11eb-9d71-11c9d606cbb1.png)